classes/steps/actions/webservice_action_step.php

```diff
@@ -0,0 +1,162 @@
+    }
+
+    /**
+     * Runs the configured step.
+     *
+     * @param $trigger
+     * @param $event
+     * @param $stepresults - result of previousstep to include in processing this step.
+     * @return array if execution was succesful and the response from the execution.
+     */
+    public function execute($step, $trigger, $event, $stepresults) {
+        global $CFG, $DB;
+        require_once($CFG->dirroot . '/admin/tool/trigger/guzzle/autoloader.php');
+
+        $this->update_datafields($event, $stepresults);
+
+        // Passing any data from previous steps through by applying template magic.
+        $username = $this->render_datafields($this->username);
+        $functionname = $this->render_datafields($this->functionname);
+        $params = $this->render_datafields($this->params);
+
+        // Assume the role of the provided user given their {username}.
+        $user = $DB->get_record('user', ['username' => $username], '*', MUST_EXIST);
+        $user->ignoresesskey = true; // This bypasses the sesskey check for the external api call.
+        \core\session\manager::set_user($user);
+        set_login_session_preferences();
+
+        // Execute the provided function name passing with the given parameters.
+        $response = \external_api::call_external_function($functionname, json_decode($params, true));
+        if ($response['error']) {
+            return [false, $response['exception']];
+        }
+
+        $stepresults = $response;
+        return [true, $stepresults];
+    }
+
+    /**
+     * {@inheritDoc}
+     * @see \tool_trigger\steps\base\base_step::add_extra_form_fields()
+     */
+    public function form_definition_extra($form, $mform, $customdata) {
+        // URL.
+        $attributes = ['size' => '50', 'placeholder' => 'my_function_name'];
+        $mform->addElement('text', 'functionname', get_string('webserviceactionfunctionname', 'tool_trigger'), $attributes);
+        $mform->setType('functionname', PARAM_RAW_TRIMMED);
+        $mform->addRule('functionname', get_string('required'), 'required');
+        $mform->addHelpButton('functionname', 'webserviceactionfunctionname', 'tool_trigger');
+
+        // Who.
+        $attributes = ['placeholder' => 'username'];
+        $mform->addElement('text', 'username', get_string('webserviceactionusername', 'tool_trigger'), $attributes);
+        $mform->setType('username', PARAM_ALPHANUMEXT);
+        $mform->addRule('username', get_string('required'), 'required');
+        $mform->setDefault('username', 'admin'); // Set the default to admin?
+        $mform->addHelpButton('username', 'webserviceactionusername', 'tool_trigger');
+
+        // Params.
+        $attributes = ['cols' => '50', 'rows' => '5'];
+        $mform->addElement('textarea', 'params', get_string('webserviceactionparams', 'tool_trigger'), $attributes);
+        $mform->setType('params', PARAM_RAW_TRIMMED);
+        $mform->addHelpButton('params', 'webserviceactionparams', 'tool_trigger');
```

Comment:
would it be too crazy to attempt to validate the shape of the inputs matches the ws inputs?

Reply:
No not too crazy. Was wondering if we just let it explode on run since it could be any shape, but maybe it would be better to validate it on save. I think I've found something which might help us do just that so will try it out.